labs/b4.md

```diff
@@ -0,0 +1,269 @@
+so you have to compile those yourself in order to compile what’s at hand. Most
+of the time, these steps are simplified through the use of a Makefile, which
+controls the Make build system.
+
+In most source tarballs, there is usually a Makefile that contains a set of
+directives to compile a project. This is because there are usually multiple
+files across many directories that need to be compiled together into the
+final executable. On top of that, there are multiple settings that control, for
+example, optimizations, size of the executable, static vs. dynamic linking,
+and whether to link against system libraries or alternatives.
+
+Many projects that have to be compiled are usually in C, C++, or similar
+lower-level languages. On most Linux distributions, there are usually three
+compiler “options”. There is the GNU Toolchain which provides `gcc` or the GNU
+C Compiler, LLVM which provides `clang`, and Intel’s proprietary toolkit,
+`icc`. Both `gcc` and `clang` are open-source and free software. While `clang`
+was designed to be a drop-in replacement for `gcc`, several quirks remain that
+make porting large programs nontrivial.
+
+To compile software that provides a Makefile, assuming you have the
+dependencies, simply type:
+
+    $ make
+
+This is will usually choose the correct compiler and compile the whole project.
+Once the compilation is done, the resulting executables are usually stored in
+`./bin` or `./build`.
+
+GCC and clang also have compiler flags that allow certain features to be
+enabled. Usually the flags that actually matter are optimization flags.
+Depending on what you want to optimize for, either space or memory or speed,
+there is a flag for it.
+
+Now, we will make a very simple application in C that prints “Hello Penguin!”
+named `hellopenguin`. Run:
+
+    $ nano hellopenguin.c
+
+to create a file named `hellopenguin.c`, and type in the following:
+
+    #include <stdio.h>
+    int main()
+    {
+        printf("Hello Penguin!");
+        return 0;
+    }
+
+Now save and exit.
+
+We will now compile the source file that you have just written:
+
+    $ gcc hellopenguin.c -o hellopenguin
+
+What this does is take the source file `hellopenguin.c` and compile it, writing
+the executable output to a file named `hellopenguin`.
+
+# How do I package stuff?
+
+Packaging manually for Debian can frustrating, especially for first timers.
+That’s why for this class, we’ll be using a really cool Ruby gem called `fpm` which
+simplifies the task of packaging a lot. Please note that if you were going to
+package for Debian’s own repositories, this is NOT the  correct/formal way to
+go about packaging. It is merely a great way to backport or package your own
+applications extremely quickly. First, make sure `ruby` and its own package
+manager called `gem` is installed. If they aren't, run `sudo apt install
+ruby-dev ruby-ffi`.  Now run the following to install `fpm` locally:
```

Comment:
These instructions are redundant from above. The first one should be removed as the command makes more sense in this context.